File: frontend/src/services/patient.service.ts
// Patient Service
// Acts as a proxy layer between components and repositories

import type { IPatientRepository } from "@/repositories";
import type {
  PaginatedResponse,
  PatientListDTO,
  PatientDetailDTO,
  PatientSummaryDTO,
  PatientSearchParams,
  CreatePatientRequest,
  VitalSignDTO,
  VitalChartDataDTO,
  ConditionDTO,
  MedicationDTO,
  LabPanelDTO,
  TimelineEventDTO,
  CreateConditionRequest,
  CreateMedicationRequest,
  RecordVitalsRequest,
  OrderLabsRequest,
  LabOrderDTO,
  RecordVitalsResponse,
  ObservationSearchParams,
  ConditionSearchParams,
  MedicationSearchParams,
  ObservationListDTO,
  ConditionListDTO,
  MedicationListDTO,
} from "@/types";

export class PatientService {
  constructor(private repository: IPatientRepository) {}

  /**
   * Get paginated list of patients with optional filters
   */
  async getPatients(
    params: PatientSearchParams = {}
  ): Promise<PaginatedResponse<PatientListDTO>> {
    return this.repository.getAll(params);
  }

  /**
   * Get detailed patient information by ID
   */
  async getPatient(id: string): Promise<PatientDetailDTO> {
    return this.repository.getById(id);
  }

  /**
   * Get patient summaries for dashboard/quick lists
   */
  async getPatientSummaries(limit?: number): Promise<PatientSummaryDTO[]> {
    return this.repository.getSummaries(limit);
  }

  /**
   * Get patient vital signs
   */
  async getVitals(patientId: string): Promise<VitalSignDTO[]> {
    return this.repository.getVitals(patientId);
  }

  /**
   * Get patient vitals chart data
   */
  async getVitalsChart(patientId: string): Promise<VitalChartDataDTO[]> {
    return this.repository.getVitalsChart(patientId);
  }

  /**
   * Get patient conditions
   */
  async getConditions(
    patientId: string,
    includeResolved: boolean = false
  ): Promise<ConditionDTO[]> {
    return this.repository.getConditions(patientId, includeResolved);
  }

  /**
   * Get patient medications
   */
  async getMedications(
    patientId: string,
    includeDiscontinued: boolean = false
  ): Promise<MedicationDTO[]> {
    return this.repository.getMedications(patientId, includeDiscontinued);
  }

  /**
   * Get patient lab results
   */
  async getLabPanels(patientId: string): Promise<LabPanelDTO[]> {
    return this.repository.getLabPanels(patientId);
  }

  /**
   * Get patient timeline events
   */
  async getTimeline(patientId: string): Promise<TimelineEventDTO[]> {
    return this.repository.getTimeline(patientId);
  }

  /**
   * Create a new condition for a patient
   */
  async createCondition(
    patientId: string,
    request: CreateConditionRequest
  ): Promise<ConditionDTO> {
    return this.repository.createCondition(patientId, request);
  }

  /**
   * Create a new medication for a patient
   */
  async createMedication(
    patientId: string,
    request: CreateMedicationRequest
  ): Promise<MedicationDTO> {
    return this.repository.createMedication(patientId, request);
  }

  /**
   * Record vital signs for a patient
   */
  async recordVitals(
    patientId: string,
    request: RecordVitalsRequest
  ): Promise<RecordVitalsResponse> {
    return this.repository.recordVitals(patientId, request);
  }

  /**
   * Order labs for a patient
   */
  async orderLabs(
    patientId: string,
    request: OrderLabsRequest
  ): Promise<LabOrderDTO> {
    return this.repository.orderLabs(patientId, request);
  }

  /**
   * Create a new patient
   */
  async createPatient(
    request: CreatePatientRequest
  ): Promise<PatientDetailDTO> {
    return this.repository.createPatient(request);
  }

  /**
   * Get cross-patient observations list
   */
  async getAllObservations(
    params: ObservationSearchParams
  ): Promise<PaginatedResponse<ObservationListDTO>> {
    return this.repository.getAllObservations(params);
  }

  /**
   * Get cross-patient conditions list
   */
  async getAllConditions(
    params: ConditionSearchParams
  ): Promise<PaginatedResponse<ConditionListDTO>> {
    return this.repository.getAllConditions(params);
  }

  /**
   * Get cross-patient medications list
   */
  async getAllMedications(
    params: MedicationSearchParams
  ): Promise<PaginatedResponse<MedicationListDTO>> {
    return this.repository.getAllMedications(params);
  }
}

File: FhirHubServer/src/FhirHubServer.Api/Features/PatientManagement/Services/PatientService.cs
using FhirHubServer.Api.Common.DependencyInjection;
using FhirHubServer.Api.Common.DTOs;
using FhirHubServer.Api.Features.PatientManagement.DTOs;
using FhirHubServer.Api.Features.PatientManagement.Repositories;

namespace FhirHubServer.Api.Features.PatientManagement.Services;

public class PatientService : IPatientService, IScopedService
{
    private readonly IPatientRepository _repository;

    public PatientService(IPatientRepository repository)
    {
        _repository = repository;
    }

    public Task<PaginatedResponse<PatientListDto>> GetAllAsync(PatientSearchParams searchParams, CancellationToken ct = default)
        => _repository.GetAllAsync(searchParams, ct);

    public Task<PatientDetailDto?> GetByIdAsync(string id, CancellationToken ct = default)
        => _repository.GetByIdAsync(id, ct);

    public Task<IEnumerable<PatientSummaryDto>> GetSummariesAsync(int limit, CancellationToken ct = default)
        => _repository.GetSummariesAsync(limit, ct);

    public Task<IEnumerable<VitalSignDto>> GetVitalsAsync(string patientId, CancellationToken ct = default)
        => _repository.GetVitalsAsync(patientId, ct);

    public Task<IEnumerable<VitalChartDataDto>> GetVitalsChartAsync(string patientId, CancellationToken ct = default)
        => _repository.GetVitalsChartAsync(patientId, ct);

    public Task<IEnumerable<ConditionDto>> GetConditionsAsync(string patientId, bool includeResolved, CancellationToken ct = default)
        => _repository.GetConditionsAsync(patientId, includeResolved, ct);

    public Task<IEnumerable<MedicationDto>> GetMedicationsAsync(string patientId, bool includeDiscontinued, CancellationToken ct = default)
        => _repository.GetMedicationsAsync(patientId, includeDiscontinued, ct);

    public Task<IEnumerable<LabPanelDto>> GetLabPanelsAsync(string patientId, CancellationToken ct = default)
        => _repository.GetLabPanelsAsync(patientId, ct);

    public Task<IEnumerable<TimelineEventDto>> GetTimelineAsync(string patientId, CancellationToken ct = default)
        => _repository.GetTimelineAsync(patientId, ct);

    // Write operations
    public Task<PatientDetailDto> CreatePatientAsync(CreatePatientRequest request, CancellationToken ct = default)
        => _repository.CreatePatientAsync(request, ct);

    public Task<ConditionDto> CreateConditionAsync(string patientId, CreateConditionRequest request, CancellationToken ct = default)
        => _repository.CreateConditionAsync(patientId, request, ct);

    public Task<MedicationDto> CreateMedicationAsync(string patientId, CreateMedicationRequest request, CancellationToken ct = default)
        => _repository.CreateMedicationAsync(patientId, request, ct);

    public Task<RecordVitalsResponse> RecordVitalsAsync(string patientId, RecordVitalsRequest request, CancellationToken ct = default)
        => _repository.RecordVitalsAsync(patientId, request, ct);

    public Task<LabOrderDto> OrderLabsAsync(string patientId, OrderLabsRequest request, CancellationToken ct = default)
        => _repository.OrderLabsAsync(patientId, request, ct);

    // Cross-patient queries
    public Task<PaginatedResponse<ObservationListDto>> GetAllObservationsAsync(ObservationSearchParams searchParams, CancellationToken ct = default)
        => _repository.GetAllObservationsAsync(searchParams, ct);

    public Task<PaginatedResponse<ConditionListDto>> GetAllConditionsAsync(ConditionSearchParams searchParams, CancellationToken ct = default)
        => _repository.GetAllConditionsAsync(searchParams, ct);

    public Task<PaginatedResponse<MedicationListDto>> GetAllMedicationsAsync(MedicationSearchParams searchParams, CancellationToken ct = default)
        => _repository.GetAllMedicationsAsync(searchParams, ct);
}
